db/mysql.js

```diff
@@ -676,5 +677,90 @@ module.exports = function (log, error) {
+  //
+  // The database enforces mutual exclusion of access to events in order to
+  // prevent publication of duplicates.  If another client is currently
+  // accessing the events, then the callback is not invoked and this method
+  // will reject with an EventQueueLockedError error object.
+
+  var GET_UNPUBLISHED_EVENTS = 'CALL getUnpublishedEvents_1()'
+  var ACK_PUBLISHED_EVENTS = 'CALL ackPublishedEvents_1(?)'
+
+  function EventQueueLockedError(message) {
+    this.message = message || 'event queue locked'
+    Error.captureStackTrace(this, EventQueueLockedError)
+  }
+  util.inherits(EventQueueLockedError, Error)
+
+  MySql.prototype.processUnpublishedEvents = function (callback, errback) {
+    // Internally we use a db-level lock for mutual exclusion.
+    // We have to call GET_LOCK() and RELEASE_LOCK() on the same
+    // db connection or things will get very confused.
+    return this.getConnection('MASTER')
+      .then(
+        function (connection) {
+          // We must manually release this connection when we're done.
+          return query(connection, GET_UNPUBLISHED_EVENTS, [])
+            .then(
+              function (results) {
+                // Did we successfully acquire the lock?
+                if (!results[0][0].lockAcquired) {
+                  throw new EventQueueLockedError()
+                }
+                // We're now holding a db-level lock that must be released.
+                // Luckily MySQL will clear it if our connection dies,
+                // and we kill the connection in the event of error.
+                var events = results[2]
+                return callback(events)
+                  .then(
+                    function (numProcessed) {
+                      var ackPos
+                      // Default to acknowledging the entire list.
+                      if (typeof numProcessed === 'undefined') {
+                        numProcessed = events.length
+                      }
+                      if (numProcessed === 0) {
+                        ackPos = 0
+                      } else {
+                        ackPos = events[numProcessed - 1].pos
+                      }
+                      // ACK_PUBLISHED_EVENTS will release the lock.
+                      return query(connection, ACK_PUBLISHED_EVENTS, [ackPos])
+                    }
+                  )
+              }
+            )
+            .then(
+              function (res) {
+                // All is well, return the connection to the pool.
+                connection.release()
+                return res
+              },
+              function (err) {
+                // Something went wrong.  Unless we're sure we don't have the
+                // lock, destroy the connection so we don't leave it hanging.
+                if (err instanceof EventQueueLockedError) {
+                  connection.release()
+                } else {
+                  connection.destroy()
```

Comment:
sledgehammer approach works. maybe we ought to log this case to find problematic callbacks